lib/maths/CTreeShapFeatureImportance.cc

```diff
@@ -200,67 +213,87 @@ void CTreeShapFeatureImportance::extendPath(SPath& path,
+    int pathDepth = nextIndex - 1;
+    double nextFractionOne{scalePath[pathDepth]};
     double fractionOne{path.fractionOnes(pathIndex)};
     double fractionZero{path.fractionZeros(pathIndex)};
-
     if (fractionOne != 0) {
+        double pD = static_cast<double>(pathDepth + 1);
+        double countUp{fractionZero / pD};
+        double countDown{(pD - 1.0) * (fractionOne / pD)};
         for (int i = pathDepth - 1; i >= 0; --i) {
-            double tmp = nextFractionOne * static_cast<double>(pathDepth + 1) /
-                         (static_cast<double>(i + 1) * fractionOne +
-                          std::numeric_limits<double>::epsilon());
-            nextFractionOne = path.scale(i) - tmp * fractionZero *
-                                                  static_cast<double>(pathDepth - i) /
-                                                  static_cast<double>(pathDepth + 1);
+            double tmp = nextFractionOne / countDown;
+            nextFractionOne = scalePath[i] - tmp * countUp;
             total += tmp;
+            countUp += fractionZero / pD;
+            countDown -= (fractionOne / pD);
         }
     } else {
-        for (int i = pathDepth - 1; i >= 0; --i) {
-            total += path.scale(i) * static_cast<double>(pathDepth + 1) /
-                     (fractionZero * static_cast<double>(pathDepth - i) +
-                      std::numeric_limits<double>::epsilon());
-        }
+        double pD{static_cast<double>(pathDepth)};
+        total =
+            std::accumulate(scalePath, scalePath + pathDepth, 0.0,
+                            [&pD](double a, double b) { return a + b / pD--; });
+        total *= static_cast<double>(pathDepth + 1) / fractionZero;
     }
 
     return total;
 }
 
-void CTreeShapFeatureImportance::unwindPath(SPath& path, std::size_t pathIndex) {
-    int pathDepth = static_cast<int>(path.depth());
-    double nextFractionOne{path.scale(pathDepth)};
+void CTreeShapFeatureImportance::unwindPath(PathElementAccessor& path,
+                                            TDoubleVecIt& scalePath,
+                                            int pathIndex,
+                                            int& nextIndex) {
+    int pathDepth{nextIndex - 1};
+    double nextFractionOne{scalePath[pathDepth]};
     double fractionOne{path.fractionOnes(pathIndex)};
     double fractionZero{path.fractionZeros(pathIndex)};
 
     if (fractionOne != 0) {
-        for (int i = pathDepth - 1; i >= 0; --i) {
-            double tmp = nextFractionOne * (pathDepth + 1) /
-                         (static_cast<double>(i + 1) * fractionOne);
-            nextFractionOne = path.scale(i) - tmp * fractionZero * (pathDepth - i) /
-                                                  (pathDepth + 1);
-            path.s_Scale[i] = tmp;
+        double countUp{0.0};
+        double countDown{static_cast<double>(nextIndex)};
+        for (int i = pathDepth; i >= 0; --i, ++countUp, --countDown) {
```

Comment:
We should be able to use similar tricks to pull operations out of the loop here and below.

Reply:
I had here similar observation as above, changed it in 3276b9dd.